### PR TITLE
Favour class-based REPL wrappers

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by dwijnand** [comment]
Added a test and then a change so that my little value class test works in class-based.

File: src/repl/scala/tools/nsc/interpreter/ILoop.scala [hunk 1]
```diff
@@ -383,7 +383,7 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter) extend
       }
   }
 
-  private def pathToPhaseWrapper = intp.originalPath("$r") + ".phased.atCurrent"
```
Comment (retronym) [inline src/repl/scala/tools/nsc/interpreter/ILoop.scala]: `s/definesTopLevel/isIObjectWrapper` throughout as the policy of when to pick one wrapper style over another might change and the code using this boolean only needs to know the what, not the why.

File: src/repl/scala/tools/nsc/interpreter/IMain.scala [hunk 2]
```diff
@@ -316,14 +316,14 @@ class IMain(initialSettings: Settings, protected val out: JPrintWriter) extends
...
-  def originalPath(sym: Symbol): String  = translateOriginalPath(typerOp path sym)
+  def originalPath(name: String, definesTopLevel: Boolean): String = originalPath(TermName(name), definesTopLevel)
+  def originalPath(name: Name, definesTopLevel: Boolean): String   = translateOriginalPath(typerOp path name, definesTopLevel)
+  def originalPath(sym: Symbol, definesTopLevel: Boolean): String  = translateOriginalPath(typerOp path sym, definesTopLevel)
 
   /** For class based repl mode we use an .INSTANCE accessor. */
-  def readInstanceName = if (isClassBased) ".INSTANCE" else ""
-  def translateOriginalPath(p: String): String = {
-    if (isClassBased) p.replace(sessionNames.read, sessionNames.read + readInstanceName) else p
+  def readInstanceName(definesTopLevel: Boolean) = if (isClassBased && !definesTopLevel) ".INSTANCE" else ""
```
Comment (retronym) [inline src/repl/scala/tools/nsc/interpreter/IMain.scala]: As discussed, we might be able to add `def INSTANCE = this` to the object based wrappers and then just always emit the same path here.

**Comment by dwijnand** [comment]
> As discussed, we might be able to add `def INSTANCE = this` to the object based wrappers and then just always emit the same path here.

> `s/definesTopLevel/isIObjectWrapper`

The change to always generating and using `INSTANCE` worked out fairly smoothly (had to change some reflection code, and stumbled on a 1 line fix, please check, and there's an indentation change in a check file but it looked wrong before too).

Having done that, my `definesTopLevel` boolean ends up being defined once and used once!  So I kept it named that way.

Review away!  Also, I have a stack of other commits in here, to get things working for me.  (I'm using `brew install graalvm-ce-java11`, so both GraalVM things and Java 11 things).  Also, I pulled `INSTANCE` into `SessionNames`, separately.

File: src/repl/scala/tools/nsc/interpreter/IMain.scala [hunk 3]
```diff
@@ -1083,7 +1087,7 @@ class IMain(initialSettings: Settings, protected val out: JPrintWriter) extends
             val i =
               if (s.isModule) {
                 if (inst == null) null
-                else mirror.reflect((inst reflectModule s.asModule).instance)
+                else mirror.reflect((mirror reflectModule s.asModule).instance)
```
Comment (dwijnand) [inline src/repl/scala/tools/nsc/interpreter/IMain.scala]: Unsure scala reflection change that unbroke `run/t8935-object.scala`

File: test/files/run/repl-colon-type.check [hunk 4]
```diff
@@ -39,7 +39,7 @@ scala> :type protected lazy val f = 5
  enclosing object $eval in package $line13 is not a subclass of
  object $iw where target is defined
          lazy val $result = f
-                                                  ^
+                                                           ^
```
Comment (dwijnand) [inline test/files/run/repl-colon-type.check]: indentation change that was broken before too

**Comment by retronym** [comment]
Pushed a commit to make test/files/run/repl-serialization.scala work again. As per my hunch, we mustn't fall back to class-based wrappers for lines that only define classes.

Why? If the class `C` with a body references a value `x` from a previous line, Spark needs that reference to be indirected through a captured `val line$2$read = $line3.$read.INSTANCE; import line$2$read.x`   . This way a subsequent line defining `() => new C()` will serialize an object graph reaching all the way to the the value of `x`.

**Comment by dwijnand** [comment]
@retronym I had a real go at trying to implement your idea of tacking on a "defining macro implementations is unsupported, restart the REPL w/ .. setting" error message, but to me it seemed like there wasn't any pre-existing place in the REPL where this should happen.  The closest is the class/companion object messaging, but that's on positive compilation.

So instead I approximated support for macros when using class-based REPL wrappers.  Do you find that too risky/unprincipled/low ROI?  Feedback welcome.

**Comment by retronym** [comment]
I've got a third option: we can teach the macro engine how to navigate the `.INSTANCE.iw().iw()` methods to get to the singleton instance of the owner of a macro implementation method or bundle class. [WIP](https://github.com/retronym/scala/tree/topic/macro-class-based).

I think this is worth the extra effort to avoid the risk of the syntactic heuristic which could be triggered in a Spark REPL session that uses some other type named 'Context` in a method signature.

**Comment by dwijnand** [comment]
Well, that was simpler than I thought it would've been.

Tomorrow I'll get the test suite passing with it defaulted to true.

Obviously this is for landing the change in 2.13, not in 2.12.  Same goes with magic imports.

**Comment by dwijnand** [comment]
Next week I need to get those partest test cases not change like that...

**Comment by dwijnand** [comment]
GOTO #83

**Pull request closed without merging**